### ISSUE TITLE
nan loss for large sequence_length

### ISSUE THREAD

**Issue by horenbergerb** [horenbergerb/BernoulliDiffusion#15]
As the sequence_length gets larger, the probabilities of individual outcomes rapidly approach zero. I think this is causing nan loss to occur.

Possibly some smart clamping in math_utils would be enough to fix this. I'd like to figure out a principled solution, if possible.

**Comment by horenbergerb** [comment]
I think I got to the bottom of this with e159522d64. The issue was improper clipping in the math_utilities as well as an exploding gradient problem.
I think this is mostly solved, but the exploding/vanishing gradient issue is certainly still lurking. I patched it up with torch.nn.utils.clip_grad_norm_, so we'll see if that stands the test of time.
